### PR TITLE
Add GitHub and gitlab registration

### PR DESCRIPTION
**Pull request by mouazq**
We welcome contributions involving code, documentation, or design. If you'd like to contribute, please check out [our guide to contributing](https://github.com/ContributorCovenant/beacon/blob/release/CODE_OF_CONDUCT.md).

All contributions, including pull requests, issues, and comments, are governed by our [code of conduct](https://github.com/ContributorCovenant/beacon/blob/release/CODE_OF_CONDUCT.md).

## Problem
Integrate Github & Gitlab authentication, so it is possible to sign up, sign in and link accounts using Github and Gitlab provider.
## Solution
The solution uses omniauth gem that is integrated with `devise`.
The solution is inspired from [here](https://github.com/omniauth/omniauth/wiki/Managing-Multiple-Providers) and [here](https://github.com/virtualforce/Devise-Omniauth-Multiple-Providers)

The controller to handle the requests is in `omniauth_callabacks_controller` 
A new relation named `credential` allow to link one account with multiple identities.

## Todo
Nothing.

## Notes for Reviewers
Anything that you want to tell the people who are going to review your pull request.

## Checklist
- [ ] Reasonable and adequate test coverage
- [ ] Requires a database migration
- [ ] Any new permissions are present in `app/models/concerns/permissions.rb`
- [ ] Any new environment variables are documented and added to `.env.development.example`

### CODE REVIEW

File: spec/models/credential_spec.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+      context 'it does not have an associated account' do
+        it 'should not be valid' do
+          credential.account = nil
+          expect(credential.valid?).to eq false
+        end
+      end
+    end
+
+    context 'credential has a in the Account model configured provider' do
```
Comment (CoralineAda) [inline spec/models/credential_spec.rb]: "has a __ in the Account model"?

File: .gitignore [hunk 2]
```diff
@@ -32,3 +32,7 @@
 /config/secrets/*
 
 .env.development
+.env.test
+
+# Ignore credentials
+/config/credentials.yml.enc
```
Comment (CoralineAda) [inline .gitignore]: I think this is causing your tests to fail on Travis. Why did you add it?

Comment (mouazq) [reply .gitignore]: Somehow. I need to remove the credentials that I have locally and generate my own. When I do this, this file is being changed so I thought of ignoring it.

**Comment by CoralineAda** [comment]
Try changing it locally and not checking it in?

**Comment by mouazq** [comment]
> This looks really great, thanks so much! A few small comments and questions, but nothing major. Let me know when you resolve the merge conflicts.
> 
> Also, would you mind replacing the ENV['FOO'] calls with settings? The way it works is you read the ENV from here: https://github.com/ContributorCovenant/beacon/blob/release/config/settings/default.yml
> And then in the code you retrieve the setting like `Setting.oauth(:github_app_key)` or whatever. Let me know if that doesn't make sense to you.

Ok. 

I tried to do that but the `default.yml` file is not loaded in the `devise.rb` when starting the app and I am getting an error.

I did it this way but it didn't work:

```
  config.omniauth :github, Setting.oauth(:github_app_id),
                           Setting.oauth(:github_app_secret), scope: 'user'
  config.omniauth :gitlab, Setting.oauth(:gitlab_app_id),
                           Setting.oauth(:gitlab_app_secret), scope: 'read_user'
```

**Comment by CoralineAda** [comment]
Try changing the order of the require?
